### ISSUE TITLE
Auto-detect Tor

### ISSUE THREAD

**Issue by hudon** [hudon/refraction-hs#80]
Currently relies on user to have Tor running but the app should check for a control port first and foremost. If no tor daemon is detected then refraction is sent into a loop until tor is running

**Comment by nopara73** [comment]
Are you using socks port stream isolation or control port `NEWNYM` request to separate identities? Or Xim doesn't require such thing?

**Comment by hudon** [comment]
@nopara73 `NEWNYM` would be bad UX I think, since that would break any existing sessions the user has with other Tor connections, right?
I think stream isolation is the right way forward but this is not implemented yet (in its current form refraction piggybacks on the existing circuit)

**Comment by nopara73** [comment]
`NEWNYM` is a bad UX design depending on how often you issue it, since it is rate limited, which depending on how often you are issuing it, it can take up to 30 seconds to take effect, usually 7ish seconds. If you don't issue it too often, it'll happen within a second.  
  Also `NEWNYM` does not dispose existing connections, it simply marks them dirty, you must dispose them yourself and build new ones, which will get to new streams.  
  On the other hand working with it can be really difficult, but if you have a decent Tor library, this shouldn't be a big issue. Of course if this is the only thing you use the ControlPort for, then not using it can remove a bunch of complexity.  

  Stream isolation is superfast, and usually enough. I'd say 99.99% of the time, there are some convoluted attacks, I will link at the end to be aware of, but they seem to be fairly impossible. Btw JoinMarket uses Stream isolation.  
  Also just to add to your worry list as I noticed Tor will sometimes silently fail to isolate streams (like 10% of the times, but oh well.)  
If you choose to go down the stream isolation path and your library doesn't handle it for you out of the box, then I give you an advice that'll save you a lot of "wtf" moments. Tor's socks port stream isolation works this way: you specify a username and the password when you build a new connection, but this username password pair surprisingly has nothing to do with authentication, it simply facilitates stream isolation. So you can specify arbitrary values, eg: "foo" "fooPassword", and every time you use these values Tor puts the connection to the "foo"-"fooPassword" stream. If you do "foo"-"fooPassword2", now that's how you isolate the stream. Anyway if your library handles it for you then this information is useless, otherwise gold:)  

  Here's my investigation into Tor's source code, where I wanted to figure out if `NEWNYM` is necessary or Stream isolation is fine: https://tor.stackexchange.com/questions/16398/signal-newnym-vs-stream-isolation  
  I'd say, if this is the only thing you are using the control port for, then go with Stream isolation and don't bother with `NEWNYM`, otherwise you may want to take a look at my investigation, depending on your paranoia level.

**Comment by nopara73** [comment]
Anyway, if you have any questions regarding Tor I'm here to help, since I'm interested in this project and I am maintaining a Tor library myself.  

On a little off-topic, while we were implementing TumbleBit, we realized we fall out of the time frame of its feasibility, because of growing network fees, so at this point I feel like the only things those make sense are the most economic based mixing techniques, which is CoinJoin. If I can recall properly Xim does have the same problem, since it publishes messages to the blockchain (OP_RETURNs?) How did you justify its feasibility to revive this project? Are you not planning to apply it to Bitcoin?   

For reference here's an article of mine, where I believe I speculated quite well the future direction of Bitcoin anonymity technologies: https://medium.com/@nopara73/anonymous-bitcoin-1fae5d1e33b7
